java-util/src/main/java/io/druid/java/util/common/CompressionUtils.java

```diff
@@ -102,8 +102,13 @@ public static long zip(File directory, OutputStream out) throws IOException
         totalSize += Files.asByteSource(file).copyTo(zipOut);
       }
       zipOut.closeEntry();
-      // Workarround for http://hg.openjdk.java.net/jdk8/jdk8/jdk/rev/759aa847dcaf
+      // Workaround for http://hg.openjdk.java.net/jdk8/jdk8/jdk/rev/759aa847dcaf
       zipOut.flush();
+
+      // fsync needs to happen before the FileOutputStream is closed by the try-with-resources
+      if (out instanceof FileOutputStream) {
+        ((FileOutputStream) out).getChannel().force(true);
```

Comment:
this is weird :)
zipOut.flush() is called before which calls out.flush() which is no-op for FileOutputStream (and rightly so based on the explanation in https://stackoverflow.com/questions/31456660/fileoutputstream-does-the-close-method-calls-also-flush )
have you seen this failing in some scenario/test or are we being extra careful ?

Reply:
No I haven't seen any failures here, it was just based on Gian's suggestion to fsync before moving the file to minimize the potential for issues. Yes agreed the construct isn't great :)

Reply:
@gianm curious, have you experienced any problem that happens without this change ?

Reply:
@himanshug it varies by filesystem; some are susceptible to a problem where write-then-rename-then-crash will get you mysterious zero-length files, where you once had perfectly good normal files. One example is early versions of ext4. Another is (I think) still current versions of xfs, see this thread: https://marc.info/?l=linux-xfs&m=139864441911660&w=2

See https://www.kernel.org/doc/Documentation/filesystems/ext4.txt for some commentary in the ext4 docs. You can see how the superior kernel programmers feel about us poor, dumb application programmers :)

```
auto_da_alloc(*)	Many broken applications don't use fsync() when 
noauto_da_alloc		replacing existing files via patterns such as
			fd = open("foo.new")/write(fd,..)/close(fd)/
			rename("foo.new", "foo"), or worse yet,
			fd = open("foo", O_TRUNC)/write(fd,..)/close(fd).
			If auto_da_alloc is enabled, ext4 will detect
			the replace-via-rename and replace-via-truncate
			patterns and force that any delayed allocation
			blocks are allocated such that at the next
			journal commit, in the default data=ordered
			mode, the data blocks of the new file are forced
			to disk before the rename() operation is
			committed.  This provides roughly the same level
			of guarantees as ext3, and avoids the
			"zero-length" problem that can happen when a
			system crashes before the delayed allocation
			blocks are forced to disk.
```

In this case, where the local deep storage pusher is probably writing to a network mounted filesystem, it's not really relevant what local fses like ext4 and xfs do, but I think it's still good to be safe given that people writing filesystem drivers have opinions like the above.

Reply:
@gianm thanks for the research. ext4 doc gives some hope that it is automatically trying to take care of not doing fsync() scenario. but for NFS, who knows.
it would be good to leave a comment in the code to this thread.